[BUG] linking fails on `cargo test -p iroha_executor --features debug`

### OS and Environment

macOS M1, happens in CI also (Ubuntu)

### GIT commit hash

started in 7a41ae8, still going as of 86d3cbc

### Minimum working example / Steps to reproduce

`cargo test -p iroha_executor --features debug`

### Actual result

```
error: linking with `cc` failed: exit status: 1
  = note: Undefined symbols for architecture arm64:
            "_dbg", referenced from:
                iroha_smart_contract_utils::debug::dbg::h090653f2ad4ed03b in libiroha_smart_contract-dd8fcf4b8ebee85c.rlib[15](iroha_smart_contract-dd8fcf4b8ebee85c.2kke7rkpsclar55l.rcgu.o)
          ld: symbol(s) not found for architecture arm64
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

### Expected result

Builds and runs tests normally

### Logs

Output from one of the failed jobs: https://github.com/hyperledger/iroha/actions/runs/9481834871/job/26125652260#step:5:1050

### Who can help to reproduce?

@mversic

### Notes

_No response_

--- a/.github/workflows/iroha2-dev-pr-wasm.yaml
+++ b/.github/workflows/iroha2-dev-pr-wasm.yaml
@@ -33,4 +33,4 @@ jobs:
         run: cargo install --path tools/wasm_test_runner
       - name: Run smart contract tests on WebAssembly VM
         working-directory: smart_contract
-        run: mold --run cargo test --release --tests --target wasm32-unknown-unknown --no-fail-fast --quiet
+        run: mold --run cargo test -p iroha_smart_contract -p iroha_smart_contract_utils --release --tests --target wasm32-unknown-unknown --no-fail-fast --quiet

--- a/smart_contract/utils/Cargo.toml
+++ b/smart_contract/utils/Cargo.toml
@@ -14,6 +14,7 @@ workspace = true
 debug = []
 
 [dependencies]
+cfg-if.workspace = true
 iroha_data_model.workspace = true
 
 parity-scale-codec.workspace = true

--- a/smart_contract/utils/src/debug.rs
+++ b/smart_contract/utils/src/debug.rs
@@ -4,6 +4,9 @@
 use alloc::format;
 use core::fmt::Debug;
 
+use cfg_if::cfg_if;
+
+#[cfg(target_family = "wasm")]
 #[cfg(not(test))]
 mod host {
     #[cfg(feature = "debug")]
@@ -23,18 +26,23 @@ mod host {
 /// Print `obj` in debug representation to the stdout.
 ///
 /// Do nothing if `debug` feature is not specified
-pub fn dbg<T: Debug + ?Sized>(_obj: &T) {
-    #[cfg(feature = "debug")]
-    {
-        #[cfg(not(test))]
-        use host::dbg as host_dbg;
-        #[cfg(test)]
-        use tests::_dbg_mock as host_dbg;
+#[allow(unused_variables)]
+pub fn dbg<T: Debug + ?Sized>(obj: &T) {
+    cfg_if! {
+        if #[cfg(not(target_family = "wasm"))] {
+            // when not on wasm - just print it
+            eprintln!("dbg: {obj:?}");
+        } else if #[cfg(feature = "debug")] {
+            // when `debug` feature is enabled - call the host `dbg` function
+            #[cfg(not(test))]
+            use host::dbg as host_dbg;
+            #[cfg(test)]
+            use tests::_dbg_mock as host_dbg;
 
-        #[allow(clippy::used_underscore_binding)]
-        let s = format!("{_obj:?}");
-        // Safety: `host_dbg` doesn't take ownership of it's pointer parameter
-        unsafe { crate::encode_and_execute(&s, host_dbg) }
+            let s = format!("{obj:?}");
+            // Safety: `host_dbg` doesn't take ownership of it's pointer parameter
+            unsafe { crate::encode_and_execute(&s, host_dbg) }
+        }
     }
 }
 

--- a/smart_contract/utils/src/debug.rs
+++ b/smart_contract/utils/src/debug.rs
@@ -23,9 +23,12 @@ mod host {
     }
 }
 
-/// Print `obj` in debug representation to the stdout.
+/// Print `obj` in debug representation.
 ///
-/// Do nothing if `debug` feature is not specified
+/// When running as a wasm smart contract, prints to host's stdout.
+/// Does nothing unless `debug` feature is enabled.
+///
+/// When running outside of wasm, always prints the output to stderr
 #[allow(unused_variables)]
 pub fn dbg<T: Debug + ?Sized>(obj: &T) {
     cfg_if! {

--- a/smart_contract/utils/src/lib.rs
+++ b/smart_contract/utils/src/lib.rs
@@ -1,5 +1,7 @@
 //! Crate with utilities for implementing smart contract FFI
-#![no_std]
+// do not use `no_std` when not running in wasm
+// this is useful to implement `dbg` and `log` functions for host-side tests
+#![cfg_attr(target_family = "wasm", no_std)]
 #![allow(unsafe_code)]
 
 extern crate alloc;

--- a/smart_contract/utils/src/log.rs
+++ b/smart_contract/utils/src/log.rs
@@ -1,9 +1,11 @@
 //! WASM logging utilities
 
+use cfg_if::cfg_if;
 pub use iroha_data_model::Level;
 
 use super::*;
 
+#[cfg(target_family = "wasm")]
 #[cfg(not(test))]
 mod host {
     #[link(wasm_import_module = "iroha")]
@@ -19,19 +21,26 @@ mod host {
 
 /// Log `obj` with desired log level
 pub fn log<T: alloc::string::ToString + ?Sized>(log_level: Level, obj: &T) {
-    #[cfg(not(test))]
-    use host::log as host_log;
-    #[cfg(test)]
-    use tests::_log_mock as host_log;
+    cfg_if! {
+        if #[cfg(not(target_family = "wasm"))] {
+            // when not on wasm - just print it
+            eprintln!("{}: {}", log_level, obj.to_string());
+        } else {
+            #[cfg(not(test))]
+            use host::log as host_log;
+            #[cfg(test)]
+            use tests::_log_mock as host_log;
 
-    let log_level_id = log_level as u8;
+            let log_level_id = log_level as u8;
 
-    let msg = obj.to_string();
-    let bytes = (log_level_id, msg).encode();
-    let ptr = bytes.as_ptr();
-    let len = bytes.len();
+            let msg = obj.to_string();
+            let bytes = (log_level_id, msg).encode();
+            let ptr = bytes.as_ptr();
+            let len = bytes.len();
 
-    unsafe { host_log(ptr, len) }
+            unsafe { host_log(ptr, len) }
+        }
+    }
 }
 
 /// Construct a new event

--- a/smart_contract/utils/src/log.rs
+++ b/smart_contract/utils/src/log.rs
@@ -20,6 +20,11 @@ mod host {
 }
 
 /// Log `obj` with desired log level
+///
+/// When running as a wasm smart contract,
+///   prints to the host logging system with the corresponding level.
+///
+/// When running outside of wasm, prints the output along with its level to stderr
 pub fn log<T: alloc::string::ToString + ?Sized>(log_level: Level, obj: &T) {
     cfg_if! {
         if #[cfg(not(target_family = "wasm"))] {

